### ISSUE TITLE
RAND_DRBG_instantiate:error retrieving entropy

### ISSUE THREAD

**Issue by cjhust** [openssl/openssl#5787]
2018/03/29 15:32:25 [error] 382057#382057: *1655048 RAND_bytes() failed (SSL: error:2406C06E:random number generator:RAND_DRBG_instantiate:error retrieving entropy error:2406B072:random number generator:RAND_DRBG_generate:in error state error:2406C06E:random number generator:RAND_DRBG_instantiate:error retrieving entropy error:2406B072:random number generator:RAND_DRBG_generate:in error state)

openssl-1.1.1-pre2, nginx-1.13.9
after a while, the error logs  flood the screen.

nginx reload, the error logs stop, but after a while, the error logs  go on flood the screen.

**Comment by mspncp** [comment]
Thank you for reporting this issue. It looks like the entropy source is not configured properly. That's why the (re-)seeding fails and the DRBG enters an error state. On every call to RAND_byte() it attempts to recover from the error by reseeding, which is why your log gets flooded with error messages.

Could you please provide more details like the following?

- Configuration data: output of 'perl configdata.pm --dump'
- OS Name, Version, Hardware platform

**Comment by cjhust** [comment]
#perf configdata.pm --dump

Command line (with current working directory = .):

    /usr/bin/perl ./Configure linux-x86_64 --prefix=/root/marco/deps/openssl-1.1.1-v2/.openssl no-shared -g -O0 enable-tls1_3 enable-weak-ssl-ciphers -Wa,--noexecstack

Perl information:

    /usr/bin/perl
    5.10.1 for x86_64-linux-thread-multi

Enabled features:

    afalgeng
    aria
    asm
    async
    autoalginit
    autoerrinit
    bf
    blake2
    camellia
    capieng
    cast
    chacha
    cmac
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    gost
    hw(-.+)?
    idea
    makedepend
    md4
    mdc2
    multiblock
    nextprotoneg
    ocb
    ocsp
    pic
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    seed
    siphash
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    static-engine
    stdio
    tests
    threads
    tls
    ts
    ui-console
    whirlpool
    weak-ssl-ciphers
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    asan                    [default] OPENSSL_NO_ASAN
    crypto-mdebug           [default] OPENSSL_NO_CRYPTO_MDEBUG
    crypto-mdebug-backtrace [default] OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
    devcryptoeng            [default] OPENSSL_NO_DEVCRYPTOENG
    dynamic-engine          [forced]  
    ec_nistp_64_gcc_128     [default] OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                     [default] OPENSSL_NO_EGD
    external-tests          [default] OPENSSL_NO_EXTERNAL_TESTS
    fuzz-libfuzzer          [default] OPENSSL_NO_FUZZ_LIBFUZZER
    fuzz-afl                [default] OPENSSL_NO_FUZZ_AFL
    heartbeats              [default] OPENSSL_NO_HEARTBEATS
    md2                     [default] OPENSSL_NO_MD2 (skip crypto/md2)
    msan                    [default] OPENSSL_NO_MSAN
    rc5                     [default] OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                    [default] OPENSSL_NO_SCTP
    shared                  [option]  
    ssl-trace               [default] OPENSSL_NO_SSL_TRACE
    tls13downgrade          [default] OPENSSL_NO_TLS13DOWNGRADE
    ubsan                   [default] OPENSSL_NO_UBSAN
    unit-test               [default] OPENSSL_NO_UNIT_TEST
    zlib                    [default] 
    zlib-dynamic            [default] 
    ssl3                    [default] OPENSSL_NO_SSL3
    ssl3-method             [default] OPENSSL_NO_SSL3_METHOD

Config target attributes:

    aes_asm_src => "aes-x86_64.s vpaes-x86_64.s bsaes-x86_64.s aesni-x86_64.s aesni-sha1-x86_64.s aesni-sha256-x86_64.s aesni-mb-x86_64.s",
    aes_obj => "aes-x86_64.o vpaes-x86_64.o bsaes-x86_64.o aesni-x86_64.o aesni-sha1-x86_64.o aesni-sha256-x86_64.o aesni-mb-x86_64.o",
    apps_aux_src => "",
    apps_init_src => "",
    apps_obj => "",
    ar => "ar",
    arflags => "r",
    bf_asm_src => "bf_enc.c",
    bf_obj => "bf_enc.o",
    bn_asm_src => "asm/x86_64-gcc.c x86_64-mont.s x86_64-mont5.s x86_64-gf2m.s rsaz_exp.c rsaz-x86_64.s rsaz-avx2.s",
    bn_obj => "asm/x86_64-gcc.o x86_64-mont.o x86_64-mont5.o x86_64-gf2m.o rsaz_exp.o rsaz-x86_64.o rsaz-avx2.o",
    bn_ops => "SIXTY_FOUR_BIT_LONG",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cast_asm_src => "c_enc.c",
    cast_obj => "c_enc.o",
    cc => "gcc",
    cflags => "-Wall -O3 -pthread -m64",
    chacha_asm_src => "chacha-x86_64.s",
    chacha_obj => "chacha-x86_64.o",
    cmll_asm_src => "cmll-x86_64.s cmll_misc.c",
    cmll_obj => "cmll-x86_64.o cmll_misc.o",
    cppflags => "-DOPENSSL_USE_NODELETE -DL_ENDIAN",
    cpuid_asm_src => "x86_64cpuid.s",
    cpuid_obj => "x86_64cpuid.o",
    cxx => "g++",
    cxxflags => "-std=c++11 -Wall -O3 -pthread",
    defines => [  ],
    des_asm_src => "des_enc.c fcrypt_b.c",
    des_obj => "des_enc.o fcrypt_b.o",
    disable => [  ],
    dso_cflags => "-fPIC",
    dso_cxxflags => "",
    dso_extension => ".so",
    dso_lflags => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    dso_scheme => "DLFCN",
    ec_asm_src => "ecp_nistz256.c ecp_nistz256-x86_64.s x25519-x86_64.s",
    ec_obj => "ecp_nistz256.o ecp_nistz256-x86_64.o x25519-x86_64.o",
    enable => [  ],
    ex_libs => "-ldl -pthread",
    exe_extension => "",
    hashbangperl => "/usr/bin/env perl",
    includes => [  ],
    lflags => "",
    md5_asm_src => "md5-x86_64.s",
    md5_obj => "md5-x86_64.o",
    modes_asm_src => "ghash-x86_64.s aesni-gcm-x86_64.s",
    modes_obj => "ghash-x86_64.o aesni-gcm-x86_64.o",
    multilib => "64",
    padlock_asm_src => "e_padlock-x86_64.s",
    padlock_obj => "e_padlock-x86_64.o",
    perlasm_scheme => "elf",
    poly1305_asm_src => "poly1305-x86_64.s",
    poly1305_obj => "poly1305-x86_64.o",
    ranlib => "ranlib",
    rc => "windres",
    rc4_asm_src => "rc4-x86_64.s rc4-md5-x86_64.s",
    rc4_obj => "rc4-x86_64.o rc4-md5-x86_64.o",
    rc5_asm_src => "rc5_enc.c",
    rc5_obj => "rc5_enc.o",
    rmd160_asm_src => "",
    rmd160_obj => "",
    sha1_asm_src => "sha1-x86_64.s sha256-x86_64.s sha512-x86_64.s sha1-mb-x86_64.s sha256-mb-x86_64.s",
    sha1_obj => "sha1-x86_64.o sha256-x86_64.o sha512-x86_64.o sha1-mb-x86_64.o sha256-mb-x86_64.o",
    shared_cflag => "-fPIC",
    shared_defflag => "-Wl,--version-script=",
    shared_defines => [  ],
    shared_extension => ".so.\$(SHLIB_VERSION_NUMBER)",
    shared_extension_simple => ".so",
    shared_ldflag => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    shared_rcflag => "",
    shared_sonameflag => "-Wl,-soname=",
    shared_target => "linux-shared",
    thread_defines => [  ],
    thread_scheme => "pthreads",
    unistd => "<unistd.h>",
    uplink_aux_src => "",
    uplink_obj => "",
    wp_asm_src => "wp-x86_64.s",
    wp_obj => "wp-x86_64.o",

Recorded environment:

    AR = 
    ARFLAGS = 
    AS = 
    ASFLAGS = 
    BUILDFILE = 
    CC = /usr/local/ccache/bin/gcc
    CFLAGS = 
    CPP = 
    CPPDEFINES = 
    CPPFLAGS = 
    CPPINCLUDES = 
    CROSS_COMPILE = 
    CXX = /usr/local/ccache/bin/g++
    CXXFLAGS = 
    HASHBANGPERL = 
    LD = 
    LDFLAGS = 
    LDLIBS = 
    MT = 
    MTFLAGS = 
    OPENSSL_LOCAL_CONFIG_DIR = 
    PERL = 
    RANLIB = 
    RC = 
    RCFLAGS = 
    RM = 
    WINDRES = 

Makevars:

Use of uninitialized value $prefix in concatenation (.) or string at configdata.pm line 15358.
    AR              = ar
    ARFLAGS         = r
Use of uninitialized value $prefix in concatenation (.) or string at configdata.pm line 15358.
    CC              = /usr/local/ccache/bin/gcc
    CFLAGS          = -Wall -O3 -pthread -m64 -g -O0 -Wa,--noexecstack
    CPPDEFINES      = DSO_DLFCN HAVE_DLFCN_H NDEBUG OPENSSL_NO_DYNAMIC_ENGINE OPENSSL_PIC OPENSSL_CPUID_OBJ OPENSSL_IA32_SSE2 OPENSSL_BN_ASM_MONT OPENSSL_BN_ASM_MONT5 OPENSSL_BN_ASM_GF2m SHA1_ASM SHA256_ASM SHA512_ASM RC4_ASM MD5_ASM AES_ASM VPAES_ASM BSAES_ASM GHASH_ASM ECP_NISTZ256_ASM X25519_ASM PADLOCK_ASM POLY1305_ASM
    CPPFLAGS        = -DOPENSSL_USE_NODELETE -DL_ENDIAN
    CPPINCLUDES     = 
Use of uninitialized value $prefix in concatenation (.) or string at configdata.pm line 15358.
    CXX             = /usr/local/ccache/bin/g++
    CXXFLAGS        = -std=c++11 -Wall -O3 -pthread -g -O0 -Wa,--noexecstack
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         = 
    LDLIBS          = -ldl -pthread
Use of uninitialized value $prefix in concatenation (.) or string at configdata.pm line 15358.
    RANLIB          = ranlib
Use of uninitialized value $prefix in concatenation (.) or string at configdata.pm line 15358.
    RC              = windres

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/unix-Makefile.tmpl
    Configurations/common.tmpl

**Comment by cjhust** [comment]
docker env：
[root@9e93b08e7111 ~]# cat /etc/redhat-release 
CentOS release 6.9 (Final)

[root@9e93b08e7111 ~]# uname -a
Linux 9e93b08e7111 4.13.0-26-generic #29~16.04.2-Ubuntu SMP Tue Jan 9 22:00:44 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

cpu： Intel(R) Xeon(R) CPU E5-2620 v3 @ 2.40GHz

**Comment by mspncp** [comment]
I trivial question before starting: does /dev/urandom exist and are you able to read from it?

If you have the possibility to do it, it would be great if you could try the following steps to find out whether this is an openssl-only problem and what is the cause of the failure to retrieve entropy:

First, try if you can reproduce the error with the installed openssl tool:
```
openssl rand -hex 10
```

Then change to your source directory and try to reproduce it using the following command

```
~/src/openssl$ util/shlib_wrap.sh apps/openssl rand -hex 10
```

If both commands fail and if you are accustomed to using gdb, you could debug it as follows:

```
~/src/openssl$ util/shlib_wrap.sh apps/openssl rand -hex 10
```

Then set a breakpoint at `rand_pool_acquire_entropy` and try to find out why that function fails.

```
~/src/openssl$ util/shlib_wrap.sh gdb --args apps/openssl rand -hex 10
...
(gdb) start
(gdb) b rand_pool_acquire_entropy
(gdb) c
```
Hope that helps.

**Comment by mspncp** [comment]
Correction: the correct gdb invocation is in the last code block.

**Comment by mspncp** [comment]
Note: The function name might be RAND_POOL_acquire_entropy in your release, the name was converted to lowercase recently. It would be best if you would test with openssl-1.1.1-pre3 instead.

**Comment by kroeckx** [comment]
On Thu, Mar 29, 2018 at 11:20:44AM +0000, Matthias St. Pierre wrote:
> I trivial question before starting: does /dev/urandom exist and are you able to read from it?

Does nginx maybe chroot to directory that doesn't have
/dev/urandom access?

Do you have glibc >= 2.25? In that case it should use getrandom(),
the kernel is at least new enough for that.


Kurt

**Comment by cjhust** [comment]
# ls  /dev/urandom -l
crw-rw---- 1 root root 1, 9 Mar 24 14:01 /dev/urandom

# openssl rand -hex 10
536e0252ac32d1e2b129


# ldd --version
ldd (GNU libc) 2.12
Copyright (C) 2010 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

**Comment by cjhust** [comment]
nginx user nobody

**Comment by cjhust** [comment]
why the error logs flood the screen after a while?
at beginning, the system is OK.

**Comment by levitte** [comment]
With those permissions on `/dev/urandom`, it's no surprise that the used `nobody` can't read it.  Is that the only randomness source you have available?  `libcrypto` does try to read `/dev/random` and `/dev/srandom` as well...

**Comment by levitte** [comment]
Point is that on most systems, `/dev/urandom` is world readable as well, at least.  This is what it looks like on my system:

```
: ; ls -l /dev/*random*
crw-rw-rw- 1 root root 1, 8 mar  2 12:46 /dev/random
crw-rw-rw- 1 root root 1, 9 mar  2 12:46 /dev/urandom
```

**Comment by cjhust** [comment]
if there is no permission, the error logs will generate at the beginning?
but run about half an hour, the error logs generate.

**Comment by cjhust** [comment]
I will modify the permission, and try again.

**Comment by mspncp** [comment]
> if there is no permission, the error logs will generate at the beginning?

I have no experience with nginx, but one possible explanation for this could be the fact that nginx starts up as root and then  [drops its privileges](https://linux-audit.com/how-and-why-linux-daemons-drop-privileges). So in your case it would be possible that the master process starts up without error and only on thirst incoming connection the random generator fails.

**Comment by cjhust** [comment]
15:27 nginx starts.
16:04 nginx runs OK.
16:23 nginx has random number errors.

now, I change the nginx from user to root, and try again.

**Comment by mspncp** [comment]
No, don’t run nginx as root! It‘s better to give read permissions for /dev/*random.

**Comment by mspncp** [comment]
If the problem is caused by the fact that /dev/*random is not present in the changroot environment, as @kroeckx suspected, then you have two possibilities: Either ensure that nginx mounts the device in the changeroot environment, or switch to glibc >= 2.25, because getrandom() obtains randomness directly via system call and does not have the limitations of the random device.

**Comment by cjhust** [comment]
nginx user root
16:24 reload
18:17 nginx runs OK.

the reason maybe is the permission of /dev/urandom

**Comment by cjhust** [comment]
I will try chmod the /dev/urandom .

**Comment by kroeckx** [comment]
I'm wondering if we should add something to CHANGES that you can run into problems with applications that chroot on an old system. Or that we should try to fall back to doing the getrandom() call ourself instead of relying on glibc providing it.

**Comment by kroeckx** [comment]
When looking how to setup up nginx using a chroot, they always mention that it needs access to /dev/urandom, so I think in this case it's just going to be the permission problem.

**Comment by mspncp** [comment]
@jhust have you tried a recent version of master since #5910 has been merged? Can this issue be closed now?

**Comment by richsalz** [comment]
Closing this issue.

**Comment by pppaulpeter** [comment]
> # openssl rand -hex 10
> 536e0252ac32d1e2b129

**Hi ,
it seems i have the same issue when using openssl library on embeded  device, the log as following:**

**this is the log when i run openssl rand -hex 10:**
_3069628432:error:2406C06E:random number generator:RAND_DRBG_instantiate:error retrieving entropy:crypto/rand/drbg_lib.c:342:
3069628432:error:2406C06E:random number generator:RAND_DRBG_instantiate:error retrieving entropy:crypto/rand/drbg_lib.c:342:
3069628432:error:2406B072:random number generator:RAND_DRBG_generate:in error state:crypto/rand/drbg_lib.c:589:
3069628432:error:2406C06E:random number generator:RAND_DRBG_instantiate:error retrieving entropy:crypto/rand/drbg_lib.c:342:
3069628432:error:2406C06E:random number generator:RAND_DRBG_instantiate:error retrieving entropy:crypto/rand/drbg_lib.c:342:
3069628432:error:2406B072:random number generator:RAND_DRBG_generate:in error state:crypto/rand/drbg_lib.c:589:
3069628432:error:2406C06E:random number generator:RAND_DRBG_instantiate:error retrieving entropy:crypto/rand/drbg_lib.c:342:
3069628432:error:2406B072:random number generator:RAND_DRBG_generate:in error state:crypto/rand/drbg_lib.c:589:_
**this openssl version is 1.1.1 as following:**

OpenSSL> version
OpenSSL 1.1.1g  21 Apr 2020
OpenSSL>_ 
any thoughts? Thanks!

**Comment by mspncp** [comment]
The default OpenSSL random generator is a (deterministic) pseudo random generator (DRBG) which needs to be seeded with randomness (aka "entropy") in order to function correctly. By default, OpenSSL obtains this randomness automatically from the operating system. Unfortunately, embedded devices are notoriously bad at providing such 'randomness'. You need to fix this root cause of the problem, otherwise your cryptography will only provide fake security.

For more details, in particular about how to configure entropy sources, see the [RAND(7)](https://www.openssl.org/docs/man1.1.1/man7/RAND.html) and [RAND_DRBG(7)](https://www.openssl.org/docs/man1.1.1/man7/RAND_DRBG.html) manual pages.


**PS: please don't reuse closed issues. open a separate issue to report problems.**
